internal/httputil/mux.go

```diff
@@ -0,0 +1,46 @@
+package httputil
+
+import (
+	"strings"
+
+	"github.com/gorilla/mux"
+
+	"github.com/TykTechnologies/tyk/internal/maps"
+)
+
+// routeCache holds the raw routes as they are mapped to mux regular expressions.
+// e.g. `/foo` becomes `^/foo$` or similar, and parameters get matched and replaced.
+var pathRegexpCache = maps.NewStringMap()
+
+// RouteRegexString will convert a mux route url to a regular expression string.
+// The results for subsequent invocations with the same parameters are cached.
+func GetPathRegexp(pattern string) (string, error) {
+	val, ok := pathRegexpCache.Get(pattern)
+	if ok {
+		return val, nil
+	}
+
+	if IsMuxTemplate(pattern) {
+		dummyRouter := mux.NewRouter()
+		route := dummyRouter.PathPrefix(pattern)
+		result, err := route.GetPathRegexp()
+		if err != nil {
+			return "", err
+		}
+
+		pathRegexpCache.Set(pattern, result)
+		return result, nil
+	}
+
+	if strings.HasPrefix(pattern, "/") {
+		return "^" + pattern, nil
+	}
+	return "^.*" + pattern, nil
```

Comment:
So, this is the best i could do to maintain existing behaviour, PTAL @jeffy-mathew ;
Covered in tests under case 0-4; Quoting is pointless since it's RE2.

We could consider appending the final $, so our matching wouldn't be greedy, or even a more complex `/?$` to be more permissive to a trailing slash by default, but not to pick up anything under `^/user` if somebody just put "user" in the regexp. so even for `^/user/[0-9]/?$` it would more likely work as intended.